# How to pick the best strike and expiration for trading options

## Parse historic options data

## Load data into Optopsy

## Analyze strategy performance

In [1]:
# import and setup
import os
from pathlib import Path
import glob
import pandas as pd
import optopsy as op

In [2]:
files = glob.glob(str(Path(Path(os.getcwd()).parents[0], "data", os.path.join("rut-eod", "*.csv"))))
dfs = []
for fl in files:
    df = pd.read_csv(fl)
    dfs.append(df)
pd.concat(dfs).to_csv(Path(Path(os.getcwd()).parents[0], "data","rut_historic.csv"), index=False)

In [3]:
# Loading data into Optopsy
# Optopsy loads data directly from CSV which is why we saved it first. Loading in Optopsy does all the standardization and formatting for us.
rut_chains = op.csv_data(
    Path(Path(os.getcwd()).parents[0], "data","rut_historic.csv"),
    underlying_symbol=1,
    underlying_price=4,
    option_type=8,
    expiration=6,
    quote_date=0,
    strike=7,
    bid=14,
    ask=15
)

/opt/homebrew/Caskroom/miniforge/base/envs/py311/lib/python3.11/site-packages/optopsy/datafeeds.py:42: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  data["expiration"] = pd.to_datetime(data.expiration, infer_datetime_format=True)
/opt/homebrew/Caskroom/miniforge/base/envs/py311/lib/python3.11/site-packages/optopsy/datafeeds.py:43: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  data["quote_date"] = pd.to_datetime(data.quote_date, infer_datetime_format=True)


In [6]:
rut_chains.head()

underlying_symbol  underlying_price option_type expiration quote_date  strike  bid  ask
0               RUT            649.29           C 2010-05-21 2010-05-21   200.0  0.0  0.0
1               RUT            649.29           P 2010-05-21 2010-05-21   200.0  0.0  0.0
2               RUT            649.29           C 2010-05-21 2010-05-21   210.0  0.0  0.0
3               RUT            649.29           P 2010-05-21 2010-05-21   210.0  0.0  0.0
4               RUT            649.29           C 2010-05-21 2010-05-21   220.0  0.0  0.0

In [4]:
# Analyze strategies with Optopsy
op.short_calls(rut_chains).round(2)

/opt/homebrew/Caskroom/miniforge/base/envs/py311/lib/python3.11/site-packages/optopsy/core.py:66: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_dataset = data.groupby(cols)["pct_change"].describe()


dte_range  otm_pct_range   count  mean   std  min  25%  50%  75%    max
0      (0, 7]  (-0.5, -0.45]   666.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.05
1      (0, 7]  (-0.45, -0.4]   867.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.17
2      (0, 7]  (-0.4, -0.35]   975.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.19
3      (0, 7]  (-0.35, -0.3]  1278.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.20
4      (0, 7]  (-0.3, -0.25]  1667.0 -0.97  0.18 -1.0 -1.0 -1.0 -1.0   0.24
5      (0, 7]  (-0.25, -0.2]  2125.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.28
6      (0, 7]  (-0.2, -0.15]  2800.0 -0.96  0.20 -1.0 -1.0 -1.0 -1.0   0.36
7      (0, 7]  (-0.15, -0.1]  3731.0 -0.96  0.21 -1.0 -1.0 -1.0 -1.0   0.49
8      (0, 7]  (-0.1, -0.05]  5456.0 -0.95  0.22 -1.0 -1.0 -1.0 -1.0   0.92
9      (0, 7]  (-0.05, -0.0]  6893.0 -0.95  0.24 -1.0 -1.0 -1.0 -1.0   1.78
10     (0, 7]   (-0.0, 0.05]  3772.0 -0.95  0.48 -1.0 -1.0 -1.0 -1.0  19.00
11     (0, 7]    (0.05, 0.1]   265.0 -0.68  1.89 -1.0 -1.0 -1.0 -1.0  15.67
12     (0, 7]    (0.1, 0.15]    10.0 -1.00  0.00 -1.0 -1.0 -1.0 -1.0  -1.00
13    (7, 14]  (-0.5, -0.45]   534.0 -0.96  0.20 -1.0 -1.0 -1.0 -1.0   0.05
14    (7, 14]  (-0.45, -0.4]   824.0 -0.97  0.18 -1.0 -1.0 -1.0 -1.0   0.20
15    (7, 14]  (-0.4, -0.35]   960.0 -0.96  0.20 -1.0 -1.0 -1.0 -1.0   0.21
16    (7, 14]  (-0.35, -0.3]  1259.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.24
17    (7, 14]  (-0.3, -0.25]  1647.0 -0.97  0.19 -1.0 -1.0 -1.0 -1.0   0.27
18    (7, 14]  (-0.25, -0.2]  2046.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.33
19    (7, 14]  (-0.2, -0.15]  2646.0 -0.97  0.18 -1.0 -1.0 -1.0 -1.0   0.41
20    (7, 14]  (-0.15, -0.1]  3232.0 -0.97  0.19 -1.0 -1.0 -1.0 -1.0   0.56
21    (7, 14]  (-0.1, -0.05]  3956.0 -0.97  0.20 -1.0 -1.0 -1.0 -1.0   1.13
22    (7, 14]  (-0.05, -0.0]  4689.0 -0.97  0.25 -1.0 -1.0 -1.0 -1.0   3.61
23    (7, 14]   (-0.0, 0.05]  4403.0 -0.96  0.59 -1.0 -1.0 -1.0 -1.0  17.57
24    (7, 14]    (0.05, 0.1]   906.0 -0.85  1.15 -1.0 -1.0 -1.0 -1.0  15.67
25    (7, 14]    (0.1, 0.15]   140.0 -0.46  2.92 -1.0 -1.0 -1.0 -1.0  19.00
26    (7, 14]    (0.15, 0.2]    17.0 -1.00  0.00 -1.0 -1.0 -1.0 -1.0  -1.00
27   (14, 21]  (-0.5, -0.45]   442.0 -0.97  0.18 -1.0 -1.0 -1.0 -1.0   0.10
28   (14, 21]  (-0.45, -0.4]   733.0 -0.96  0.19 -1.0 -1.0 -1.0 -1.0   0.12
29   (14, 21]  (-0.4, -0.35]   919.0 -0.97  0.18 -1.0 -1.0 -1.0 -1.0   0.31
30   (14, 21]  (-0.35, -0.3]  1170.0 -0.97  0.18 -1.0 -1.0 -1.0 -1.0   0.35
31   (14, 21]  (-0.3, -0.25]  1565.0 -0.97  0.19 -1.0 -1.0 -1.0 -1.0   0.39
32   (14, 21]  (-0.25, -0.2]  1978.0 -0.97  0.19 -1.0 -1.0 -1.0 -1.0   0.45
33   (14, 21]  (-0.2, -0.15]  2531.0 -0.97  0.17 -1.0 -1.0 -1.0 -1.0   0.56
34   (14, 21]  (-0.15, -0.1]  3078.0 -0.97  0.17 -1.0 -1.0 -1.0 -1.0   0.65
35   (14, 21]  (-0.1, -0.05]  3560.0 -0.98  0.17 -1.0 -1.0 -1.0 -1.0   0.83
36   (14, 21]  (-0.05, -0.0]  4045.0 -0.98  0.17 -1.0 -1.0 -1.0 -1.0   1.04
37   (14, 21]   (-0.0, 0.05]  4282.0 -0.99  0.10 -1.0 -1.0 -1.0 -1.0   1.09
38   (14, 21]    (0.05, 0.1]  1659.0 -0.95  0.54 -1.0 -1.0 -1.0 -1.0   9.00
39   (14, 21]    (0.1, 0.15]   339.0 -0.78  1.70 -1.0 -1.0 -1.0 -1.0  15.67
40   (14, 21]    (0.15, 0.2]    65.0 -0.97  0.10 -1.0 -1.0 -1.0 -1.0  -0.60
41   (14, 21]    (0.2, 0.25]     8.0 -1.00  0.00 -1.0 -1.0 -1.0 -1.0  -1.00
42   (21, 28]  (-0.5, -0.45]   384.0 -0.97  0.19 -1.0 -1.0 -1.0 -1.0   0.13
43   (21, 28]  (-0.45, -0.4]   696.0 -0.96  0.22 -1.0 -1.0 -1.0 -1.0   0.24
44   (21, 28]  (-0.4, -0.35]   910.0 -0.97  0.20 -1.0 -1.0 -1.0 -1.0   0.27
45   (21, 28]  (-0.35, -0.3]  1143.0 -0.97  0.19 -1.0 -1.0 -1.0 -1.0   0.30
46   (21, 28]  (-0.3, -0.25]  1488.0 -0.96  0.21 -1.0 -1.0 -1.0 -1.0   0.34
47   (21, 28]  (-0.25, -0.2]  1884.0 -0.97  0.20 -1.0 -1.0 -1.0 -1.0   0.38
48   (21, 28]  (-0.2, -0.15]  2453.0 -0.96  0.21 -1.0 -1.0 -1.0 -1.0   0.49
49   (21, 28]  (-0.15, -0.1]  3006.0 -0.96  0.22 -1.0 -1.0 -1.0 -1.0   0.67
50   (21, 28]  (-0.1, -0.05]  3534.0 -0.96  0.23 -1.0 -1.0 -1.0 -1.0   1.12
51   (21, 28]  (-0.05, -0.0]  3949.0 -0.96  0.28 -1.

### Let’s break down the key output:

- dte_range: The range of days remaining until the options expire. For example, (0, 7] includes all options that will expire in 0 to 7 days based on the quote date.
- otm_pct_range: Range of how far out of the money the options are, expressed as a percentage. For example, (-0.5, -0.45] means the options are between 50% and 45% out of the money.
- count: The number of options contracts that fall within the specified dte_range and otm_pct_range. For instance, there are 666 options contracts that expire in 0 to 7 days and are 50% to 45% out of the money.
- mean: This is the average percent change of the options within the specified dte_range and otm_pct_range.

In [5]:
# Let’s analyze the performance of all short strangles that expire within 60 days, with a max days to expiration of 70, exit at 10 days to expiration, and each leg no further than 10% out of the money.
op.short_strangles(
    rut_chains, 
    dte_interval=60, 
    max_entry_dte=70, 
    exit_dte=10,
    otm_pct_interval=0.01,
    max_otm_pct=0.10
).round(2)

/opt/homebrew/Caskroom/miniforge/base/envs/py311/lib/python3.11/site-packages/optopsy/core.py:66: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_dataset = data.groupby(cols)["pct_change"].describe()


dte_range otm_pct_range_leg1 otm_pct_range_leg2   count  mean   std   min   25%   50%   75%   max
0     (0, 60]      (-0.1, -0.09]      (-0.1, -0.09]   398.0  0.24  0.21 -0.12  0.07  0.21  0.39  0.72
1     (0, 60]      (-0.1, -0.09]     (-0.09, -0.08]  2602.0  0.23  0.23 -0.22  0.04  0.21  0.40  0.77
2     (0, 60]      (-0.1, -0.09]     (-0.08, -0.07]  2857.0  0.26  0.25 -0.26  0.05  0.24  0.45  0.81
3     (0, 60]      (-0.1, -0.09]     (-0.07, -0.06]  3023.0  0.29  0.27 -0.28  0.07  0.27  0.49  0.85
4     (0, 60]      (-0.1, -0.09]     (-0.06, -0.05]  2941.0  0.32  0.28 -0.33  0.08  0.30  0.54  0.88
5     (0, 60]      (-0.1, -0.09]     (-0.05, -0.04]  2988.0  0.35  0.29 -0.37  0.11  0.35  0.60  0.90
6     (0, 60]      (-0.1, -0.09]     (-0.04, -0.03]  3070.0  0.39  0.31 -0.48  0.14  0.41  0.65  0.92
7     (0, 60]      (-0.1, -0.09]     (-0.03, -0.02]  3218.0  0.44  0.32 -0.57  0.21  0.49  0.72  0.93
8     (0, 60]      (-0.1, -0.09]     (-0.02, -0.01]  3145.0  0.48  0.33 -0.69  0.25  0.54  0.76  0.93
9     (0, 60]      (-0.1, -0.09]      (-0.01, -0.0]  3254.0  0.53  0.32 -0.88  0.35  0.61  0.79  0.94
10    (0, 60]      (-0.1, -0.09]       (-0.0, 0.01]  3275.0  0.58  0.31 -1.01  0.44  0.67  0.80  0.96
11    (0, 60]      (-0.1, -0.09]       (0.01, 0.02]  3314.0  0.61  0.30 -1.13  0.50  0.70  0.81  0.96
12    (0, 60]      (-0.1, -0.09]       (0.02, 0.03]  3272.0  0.62  0.29 -2.17  0.51  0.71  0.83  0.97
13    (0, 60]      (-0.1, -0.09]       (0.03, 0.04]  3273.0  0.61  0.35 -3.43  0.49  0.70  0.83  0.97
14    (0, 60]      (-0.1, -0.09]       (0.04, 0.05]  3028.0  0.59  0.43 -5.43  0.49  0.70  0.84  0.97
15    (0, 60]      (-0.1, -0.09]       (0.05, 0.06]  2525.0  0.63  0.35 -2.78  0.54  0.73  0.85  0.97
16    (0, 60]      (-0.1, -0.09]       (0.06, 0.07]  1898.0  0.68  0.28 -1.13  0.59  0.77  0.87  0.97
17    (0, 60]      (-0.1, -0.09]       (0.07, 0.08]  1222.0  0.73  0.24 -1.29  0.66  0.82  0.89  0.97
18    (0, 60]      (-0.1, -0.09]       (0.08, 0.09]   653.0  0.79  0.17 -0.22  0.74  0.85  0.90  0.97
19    (0, 60]     (-0.09, -0.08]     (-0.09, -0.08]   521.0  0.19  0.24 -0.26 -0.00  0.17  0.38  0.73
20    (0, 60]     (-0.09, -0.08]     (-0.08, -0.07]  3256.0  0.19  0.27 -0.36 -0.03  0.17  0.40  0.77
21    (0, 60]     (-0.09, -0.08]     (-0.07, -0.06]  3686.0  0.22  0.28 -0.38 -0.02  0.20  0.45  0.82
22    (0, 60]     (-0.09, -0.08]     (-0.06, -0.05]  3632.0  0.24  0.30 -0.44 -0.01  0.23  0.49  0.84
23    (0, 60]     (-0.09, -0.08]     (-0.05, -0.04]  3715.0  0.27  0.32 -0.51  0.01  0.27  0.54  0.88
24    (0, 60]     (-0.09, -0.08]     (-0.04, -0.03]  3724.0  0.30  0.34 -0.60  0.03  0.31  0.60  0.89
25    (0, 60]     (-0.09, -0.08]     (-0.03, -0.02]  3953.0  0.34  0.36 -0.70  0.06  0.38  0.66  0.90
26    (0, 60]     (-0.09, -0.08]     (-0.02, -0.01]  3926.0  0.37  0.38 -0.80  0.11  0.44  0.70  0.92
27    (0, 60]     (-0.09, -0.08]      (-0.01, -0.0]  3940.0  0.42  0.39 -0.95  0.18  0.51  0.73  0.94
28    (0, 60]     (-0.09, -0.08]       (-0.0, 0.01]  4069.0  0.46  0.39 -1.18  0.28  0.58  0.75  0.95
29    (0, 60]     (-0.09, -0.08]       (0.01, 0.02]  4080.0  0.51  0.37 -1.54  0.36  0.62  0.77  0.96
30    (0, 60]     (-0.09, -0.08]       (0.02, 0.03]  4009.0  0.54  0.35 -2.82  0.39  0.64  0.79  0.96
31    (0, 60]     (-0.09, -0.08]       (0.03, 0.04]  4078.0  0.56  0.37 -4.18  0.42  0.66  0.81  0.97
32    (0, 60]     (-0.09, -0.08]       (0.04, 0.05]  3770.0  0.58  0.42 -6.53  0.46  0.69  0.83  0.97
33    (0, 60]     (-0.09, -0.08]       (0.05, 0.06]  3200.0  0.63  0.36 -3.17  0.53  0.73  0.85  0.97
34    (0, 60]     (-0.09, -0.08]       (0.06, 0.07]  2485.0  0.69  0.29 -1.68  0.60  0.79  0.88  0.97
35    (0, 60]     (-0.09, -0.08]       (0.07, 0.08]  1681.0  0.75  0.23 -1.02  0.68  0.83  0.90  0.97
36    (0, 60]     (-0.09, -0.08]       (0.08, 0.09]   975.0  0.81  0.17 -0.28  0.76  0.86  0.91  0.97
37    (0, 60]     (-0.08, -0.07]     (-0.08, -0.07]   637.0  0.17  0.28 -0.43 -0.05  0.15  0.40  0.74
38    (0, 60]     (-0.08, 

There are 3,070 short strangles found between 0 and 60 days to expiration. The first leg, usually the put, groups contracts between 9% and 10% out of the money and the first leg between 3% and 4% out of the money. The mean return achieved by exiting this position at 10 days to expiration is 39%.